### PR TITLE
Create user model

### PR DESCRIPTION
**Pull request by aht-wtag**
- In the User model there will be unique username and email.
- Username, email, password can not be null
- Bio text field max length should be 150 and location length should be 50
- Everything tested with Rspec test cases

Story Link: https://trello.com/c/YoCIUtPI

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -65,3 +66,7 @@ group :test do
   gem "capybara"
   gem "selenium-webdriver"
 end
+
+
+gem 'rspec-rails', '~> 6.1.0'
+gem 'faker'
```
Comment (SKH-WTAG) [inline Gemfile]: Do we need these gems in production?

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class User < ApplicationRecord
+    has_secure_password
+    validates :username, presence: true, uniqueness: true
+    validates :email, presence: true, uniqueness: true
+    validates :password, presence: true
+    validates :bio, length: {maximum: 150}
+    validates :location, length: {maximum: 50}
```
Comment (SKH-WTAG) [inline app/models/user.rb]: Length is unnecessary. You can remove these

File: config/environments/development.rb [hunk 3]
```diff
@@ -17,6 +17,8 @@
   # Enable server timing
   config.server_timing = true
 
+
+
```
Comment (SKH-WTAG) [inline config/environments/development.rb]: Undo this unnecessary changes

File: db/migrate/20240416051242_drop_devise_tables.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+class DropDeviseTables < ActiveRecord::Migration[7.1]
```
Comment (SKH-WTAG) [inline db/migrate/20240416051242_drop_devise_tables.rb]: Why do we need this?

File: db/migrate/20240416094601_create_users.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class CreateUsers < ActiveRecord::Migration[7.1]
+  def change
+    create_table :users do |t|
+      t.string :username
+      t.string :email
```
Comment (SKH-WTAG) [inline db/migrate/20240416094601_create_users.rb]: Use `citext`

File: spec/factories/users.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+    factory :user do
+      username { Faker::Lorem.characters(number: rand(1..30)) }
+      email { Faker::Lorem.characters(number: rand(1..20)) + '@gmail.com' }
+      password { BCrypt::Password.create('password123') }
+    end
+end
```
Comment (SKH-WTAG) [inline spec/factories/users.rb]: Missing new line at the end of file

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,98 @@
+require "rails_helper"
+require 'faker'
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Use FFaker and configure it in rails/spec helper if needed

https://github.com/ffaker/ffaker

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+RSpec.describe User,type: :model do
+  user = User.create()
+  describe "validates username, password and email" do
+    it "validates presence of username, password, email" do
+        expect(user).not_to be_valid
+    end
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Use shoulda-machers https://github.com/thoughtbot/shoulda-matchers

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  #   # end
+
+  #   # it "username same" do
+  #   #   user.username = "tahsin"
+  #   #   user.email = "tahsin@gmail.com"
+  #   #   user.password = "hello world"
+  #   #   test_user.username = "tahsin"
+  #   #   expect(test_user).not_to be_valid
+  #   # end
+  # end
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Why did you comment this codes?

File: spec/support/factory_bot.rb [hunk 10]
```diff
@@ -0,0 +1,3 @@
+RSpec.configure do |config|
```
Comment (SKH-WTAG) [inline spec/support/factory_bot.rb]: No need additional file. Configure this on spec_helper file

File: test/fixtures/users.yml [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (SKH-WTAG) [inline test/fixtures/users.yml]: We are using rspec. Remove test folder

File: db/migrate/20240416051242_drop_devise_tables.rb (hunk 4 above)
Comment (aht-wtag) [reply db/migrate/20240416051242_drop_devise_tables.rb]: used Devise and made the table in previous commits, 

To roll back everything Devise did, followed their docs and this migration happened

File: Gemfile (hunk 1 above)
Comment (aht-wtag) [reply Gemfile]: fixed it by moving them under test and development group

File: config/environments/development.rb (hunk 3 above)
Comment (aht-wtag) [reply config/environments/development.rb]: removed these blank lines

File: spec/factories/users.rb (hunk 6 above)
Comment (aht-wtag) [reply spec/factories/users.rb]: added

File: test/fixtures/users.yml (hunk 11 above)
Comment (aht-wtag) [reply test/fixtures/users.yml]: Removed.

File: spec/models/user_spec.rb (hunk 9 above)
Comment (aht-wtag) [reply spec/models/user_spec.rb]: Removed

File: Gemfile (hunk 1 above)
Comment (aht-wtag) [reply Gemfile]: moved rspec to test and ffaker to test and development

File: app/models/user.rb (hunk 2 above)
Comment (aht-wtag) [reply app/models/user.rb]: removed the length part along with Rspec test part in user_spec.rb

File: db/migrate/20240416051242_drop_devise_tables.rb (hunk 4 above)
Comment (aht-wtag) [reply db/migrate/20240416051242_drop_devise_tables.rb]: deleted

File: spec/support/factory_bot.rb (hunk 10 above)
Comment (aht-wtag) [reply spec/support/factory_bot.rb]: https://stackoverflow.com/questions/25648456/how-to-add-config-include-factorybotsyntaxmethods-to-rspec-config-block-in

If I try to do this in the spec_helper this error is happening.Actually according to this post + documentation (https://github.com/thoughtbot/factory_bot/blob/main/GETTING_STARTED.md) I set it up

Comment (aht-wtag) [reply spec/support/factory_bot.rb]: fixed it

File: db/migrate/20240416094601_create_users.rb (hunk 5 above)
Comment (aht-wtag) [reply db/migrate/20240416094601_create_users.rb]: Done

File: spec/models/user_spec.rb (hunk 7 above)
Comment (aht-wtag) [reply spec/models/user_spec.rb]: done

File: spec/models/user_spec.rb (hunk 8 above)
Comment (aht-wtag) [reply spec/models/user_spec.rb]: added shoulda-matchers

File: Gemfile [hunk 12]
```diff
@@ -44,9 +44,16 @@ gem "bootsnap", require: false
 # Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
 # gem "image_processing", "~> 1.2"
 
+
+#pry
+gem 'pry', '~> 0.14.2'
```
Comment (SKH-WTAG) [inline Gemfile]: Use pry-rails gem in development and test only

File: app/models/user.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+    has_secure_password
+    validates :username, presence: true, uniqueness: true
+    validates :email, presence: true, uniqueness: true
```
Comment (SKH-WTAG) [inline app/models/user.rb]: Use regex to validate invalid email

File: db/migrate/20240416094601_create_users.rb (hunk 5 above)
Comment (SKH-WTAG) [reply db/migrate/20240416094601_create_users.rb]: I can not see any changes here.

File: db/migrate/20240421044950_change_email_to_be_citext_in_users.rb [hunk 14]
```diff
@@ -0,0 +1,11 @@
+class ChangeEmailToBeCitextInUsers < ActiveRecord::Migration[7.1]
+  def up
+    enable_extension("citext")
+    change_column :users, :email, :citext
+  end
+
+  def down
+    enable_extension("citext")
```
Comment (SKH-WTAG) [inline db/migrate/20240421044950_change_email_to_be_citext_in_users.rb]: Doing same thing in up and down

File: spec/models/user_spec.rb [hunk 15]
```diff
@@ -0,0 +1,13 @@
+require "rails_helper"
+require "spec_helper"
+
+RSpec.describe User, type: :model do
+  describe "validations of presence and uniqueness" do
+    it { is_expected.to validate_presence_of(:username) }
+    it { is_expected.to validate_presence_of(:email) }
+    it { is_expected.to validate_presence_of(:password) }
+    subject { FactoryBot.create(:user) }
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Is there any impact if we move subject to top or bottom?

File: spec/spec_helper.rb [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+# Given that it is always loaded, you are encouraged to keep this file as
+# light-weight as possible. Requiring heavyweight dependencies from this file
+# will add to the boot time of your test suite on EVERY test run, even for an
+# individual file that may not need all of that loaded. Instead, consider making
+# a separate helper file that requires the additional dependencies and performs
+# the additional setup, and require it from the spec files that actually need
+# it.
+#
+# See https://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
+require 'factory_bot'
```
Comment (SKH-WTAG) [inline spec/spec_helper.rb]: Try without requiring `factory_bot`

File: Gemfile (hunk 12 above)
Comment (aht-wtag) [reply Gemfile]: Done

File: spec/models/user_spec.rb (hunk 15 above)
Comment (aht-wtag) [reply spec/models/user_spec.rb]: no there is no specific impact even if we dont provide subject, FactoryBot creates it on its own and tests against them

Comment (aht-wtag) [reply spec/models/user_spec.rb]: but if we provide subject then other "it" blocks under the subject line runs tests on that specific subject

File: spec/spec_helper.rb (hunk 16 above)
Comment (aht-wtag) [reply spec/spec_helper.rb]: it gives uninitialized constant Factory bot error

File: db/migrate/20240421044950_change_email_to_be_citext_in_users.rb (hunk 14 above)
Comment (aht-wtag) [reply db/migrate/20240421044950_change_email_to_be_citext_in_users.rb]: fixed migrations and type changed to citext

File: spec/spec_helper.rb (hunk 16 above)
Comment (aht-wtag) [reply spec/spec_helper.rb]: made it work after reading the latest factory bot rails doc

File: app/models/user.rb (hunk 13 above)
Comment (aht-wtag) [reply app/models/user.rb]: added

File: Gemfile (hunk 12 above)
Comment (aht-wtag) [reply Gemfile]: Added pry-rails instead of pry

File: .rspec [hunk 17]
```diff
@@ -0,0 +1 @@
+--require spec_helper
```
Comment (SKH-WTAG) [inline .rspec:1]: Why did you require this?

File: Gemfile [hunk 18]
```diff
@@ -44,9 +44,14 @@ gem "bootsnap", require: false
 # Use Active Storage variants [https://guides.rubyonrails.org/active_storage_overview.html#transforming-images]
 # gem "image_processing", "~> 1.2"
 
+
+
```
Comment (SKH-WTAG) [inline Gemfile]: Remove unnecessary new lines.

File: app/models/user.rb [hunk 19]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+    has_secure_password
+    validates :username, presence: true, uniqueness: true
+    validates :email, format: { with: /\A(.+)@(.+)\z/, message: "Email invalid"  },
```
Comment (SKH-WTAG) [inline app/models/user.rb]:
```suggestion
    validates :email, format: { with: URI::MailTo::EMAIL_REGEXP },
```

File: db/migrate/20240416094601_create_users.rb [hunk 20]
```diff
@@ -0,0 +1,15 @@
+class CreateUsers < ActiveRecord::Migration[7.1]
+  def change
+    create_table :users do |t|
+      t.citext :username
+      t.citext :email
+      t.string :password_digest
+      t.text :bio
+      t.text :location
```
Comment (SKH-WTAG) [inline db/migrate/20240416094601_create_users.rb]: Make null false for required fields

File: spec/models/user_spec.rb (hunk 15 above)
Comment (SKH-WTAG) [reply spec/models/user_spec.rb]: You gave the subject on line 9. So, tests from line 6/7/8 won't use that subject?

File: spec/models/user_spec.rb [hunk 21]
```diff
@@ -0,0 +1,13 @@
+require "rails_helper"
+require "spec_helper"
+
+RSpec.describe User, type: :model do
+  describe "validations of presence and uniqueness" do
+    it { is_expected.to validate_presence_of(:username) }
+    it { is_expected.to validate_presence_of(:email) }
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb:5]: Add test for valid invalid email

File: db/schema.rb [hunk 22]
```diff
@@ -10,8 +10,21 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema[7.1].define(version: 0) do
+ActiveRecord::Schema[7.1].define(version: 2024_04_16_094601) do
   # These are extensions that must be enabled in order to support this database
+  enable_extension "citext"
```
Comment (SKH-WTAG) [inline db/schema.rb:15]: Where did you enable citext extension?

File: spec/models/user_spec.rb [hunk 23]
```diff
@@ -0,0 +1,13 @@
+require "rails_helper"
+require "spec_helper"
```
Comment (SKH-WTAG) [inline spec/models/user_spec.rb]: Why do we need to require these?

Comment (aht-wtag) [reply spec/models/user_spec.rb]: we do not need the spec_helper here because it is already required in rails_helper. but rails_helper is needed because shoulda matchers configuration is used in the user_spec so it needs to know where to look for the shoulda matchers

Comment (aht-wtag) [reply spec/models/user_spec.rb]: spec_helper removed, rails helper stays

File: .rspec (hunk 17 above)
Comment (aht-wtag) [reply .rspec:1]: it was auto generated by rspec while installation, purpose of this is to maybe directly execute commands from here instead of command line, in my case I should also add require rails helper then everything that needs rails helper and spec helper will be pre loaded

File: spec/models/user_spec.rb (hunk 23 above)
Comment (aht-wtag) [reply spec/models/user_spec.rb]: added rails helper in rspec no need of rails helper anymore. removing it too.

File: db/migrate/20240416094601_create_users.rb (hunk 20 above)
Comment (aht-wtag) [reply db/migrate/20240416094601_create_users.rb]: done

File: db/schema.rb (hunk 22 above)
Comment (aht-wtag) [reply db/schema.rb:15]: fixed it, added in the migration

File: spec/models/user_spec.rb (hunk 21 above)
Comment (aht-wtag) [reply spec/models/user_spec.rb:5]: added

Comment (aht-wtag) [reply spec/models/user_spec.rb:5]: in that block all test cases will use that subject. position does not matter inside the block, because every time there will be new subject generated for every test cases

File: app/models/user.rb (hunk 19 above)
Comment (aht-wtag) [reply app/models/user.rb]: added URI::MailTo

File: Gemfile (hunk 18 above)
Comment (aht-wtag) [reply Gemfile]: Removed

File: .rspec [hunk 24]
```diff
@@ -0,0 +1,2 @@
...
+--require rails_helper
```
Comment (SKH-WTAG) [inline .rspec]: Put newline at the end of file.

File: q [hunk 25]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (SKH-WTAG) [inline q]: Remove unnecessary file

**Review by SKH-WTAG — APPROVED** [review]

File: .rspec (hunk 24 above)
Comment (aht-wtag) [reply .rspec]: done
